server/src/main/scala/org/http4s/server/middleware/HttpMethodOverrider.scala

```diff
@@ -57,30 +104,44 @@ object HttpMethodOverrider {
               .getOrElse(Header(varyHeaderName.value, headerName.value))
 
           resp.withHeaders(resp.headers.put(updatedVaryHeader))
-        case QueryOverrideStrategy(_) => resp
+        case _ => resp
       }
     }
 
-    def updateRequestWithMethod(req: Request[F], om: Method): Request[F] = {
+    def updateRequestWithMethod(req: Request[G], om: Method): Request[G] = {
       val attrs = req.attributes ++ Seq(overriddenMethodAttrKey(req.method))
       req.withAttributes(attrs).withMethod(om)
     }
 
-    def ignoresOverrideIfNotAllowed(req: Request[F]): Option[Unit] =
-      if (config.overridableMethods.contains(req.method)) Some(()) else None
-
-    def getUnsafeOverrideMethod(req: Request[F]): Option[String] =
+    def getUnsafeOverrideMethod(req: Request[G]): F[Option[String]] =
       config.overrideStrategy match {
-        case HeaderOverrideStrategy(headerName) => req.headers.get(headerName).map(_.value)
-        case QueryOverrideStrategy(parameter) => req.params.get(parameter)
+        case HeaderOverrideStrategy(headerName) => F.pure(req.headers.get(headerName).map(_.value))
+        case QueryOverrideStrategy(parameter) => F.pure(req.params.get(parameter))
+        case FormOverrideStrategy(field, f) if runtimeTypeNT == typeOf[G ~> F] =>
```

Comment:
Yeah, it scared me too :S and I hate reflection. I spend ~ two hours looking for the proper/elegant way of overcoming type erasure in this kind of scenarios and there seems to a general consensus about using the following approaches:

1. Using `ClassTag[_]`, `TypeTag[_]` or `WeakTypeTag[_]` 
2. Using `Manifest[_]` 

Both are pretty similar. I had a slightly bigger preference for the first one and since [`Headerkey`](https://github.com/http4s/http4s/blob/a53dbb1747b69cbe67e637536f73d7d369d62896/core/src/main/scala/org/http4s/HeaderKey.scala#L84) is already using that approach with `ClassTag`, I went for it.

Let me know if have a better approach. I keep looking for better ways to handle this

Reply:
BTW, yes, I'd be trading `ClassCastException` for `MatchError`. It kind of feels more logical to me since we trying to match strategies. WDYT?

Reply:
I can't think of a better solution, and I think it would also be difficult to use it incorrectly.  I'm inclined to roll with it.  I'd rather have the feature, and if someone can think of something more clever later, so much the better.